### ISSUE TITLE
[Regression 1.98.1] needless_borrows_for_generic_args hangs on Option::filter with a recursive enum

### ISSUE THREAD

**Issue by roji** [rust-lang/rust-clippy#17717]
### Summary

On Rust/Clippy 1.98.1, `needless_borrows_for_generic_args` appears to hang on a borrowed predicate passed to `Option::filter` when the option contains a recursive enum backed by `Vec`. A dependency-free reproducer remains running without diagnostics after 30 seconds and was terminated. The same source completes quickly with Clippy 1.95.0 and with plain rustc 1.98.1.

Removing the unnecessary borrow, or disabling only this lint, avoids the problem. This report describes an apparent hang/severe performance regression; I have not established that execution is literally infinite.

I searched existing open and closed Clippy issues for the lint name, hangs, `HostEffectPredicate`, `Destruct`, and recursive drop/filter combinations without finding a matching report. #14902 is a broader instantiation-performance discussion; #17321 concerns a different trait-selection ICE involving async/opaque stream types.

### Minimal reproducer

Save as `repro.rs`:

```rust
#![recursion_limit = "512"]

pub enum Value {
    Array(Vec<Value>),
    Null,
}

pub fn repro(value: Option<Value>) -> Option<Value> {
    let predicate = |_: &Value| true;
    value.filter(&predicate)
}
```

There are no dependencies, procedural macros, closure captures, async functions, or unstable features. The explicit recursion limit is retained from the reproducing configuration; I have not characterized how changing it affects the problem.

Run this with a **30-second timeout** (or terminate it manually if it does not finish):

```sh
rustup run 1.98.1 clippy-driver --edition=2024 repro.rs \
  --crate-type lib --emit metadata -o repro.rmeta \
  -W clippy::needless_borrows_for_generic_args
```

For example, on systems with GNU coreutils, prefix the command with `timeout 30s` (`gtimeout 30s` on macOS with coreutils installed).

### Expected behavior

Clippy should finish promptly, with or without a diagnostic as appropriate, rather than spending tens of seconds on this tiny input.

### Actual behavior and controls

Repeated the following comparisons on the same aarch64 macOS host. These are individual wall-clock observations, not statistical benchmarks:

| Configuration | Result |
|---|---|
| Clippy 1.98.1, original reproducer | Still running without diagnostics after 30 seconds; terminated |
| Clippy 1.95.0, original reproducer | Exit 0, approximately 0.35 seconds |
| Plain rustc 1.98.1, original reproducer | Exit 0, approximately 0.18 seconds |
| Clippy 1.98.1, `value.filter(predicate)` (remove `&`) | Exit 0, approximately 0.05 seconds |
| Clippy 1.98.1, original source with only this lint allowed | Exit 0, approximately 0.05 seconds |

Commands for the original-source controls:

```sh
rustup run 1.95.0 clippy-driver --edition=2024 repro.rs \
  --crate-type lib --emit metadata -o repro195.rmeta \
  -W clippy::needless_borrows_for_generic_args

rustup run 1.98.1 rustc --edition=2024 repro.rs \
  --crate-type lib --emit metadata -o repro-rustc.rmeta

rustup run 1.98.1 clippy-driver --edition=2024 repro.rs \
  --crate-type lib --emit metadata -o repro-allowed.rmeta \
  -A clippy::needless_borrows_for_generic_args
```

For the other workaround, change only `value.filter(&predicate)` to `value.filter(predicate)` and rerun the original Clippy command.

Additional reduction observations: analogous `Option<String>::filter(&predicate)` examples completed quickly, including predicates capturing a borrowed `String` or `Vec<String>`. The recursive enum reproduces with no captures at all.

### Possible mechanism (hypothesis, not a bisected root cause)

The lint [evaluates instantiated callee predicates](https://github.com/rust-lang/rust-clippy/blob/rust-1.98.0/clippy_lints/src/needless_borrows_for_generic_args.rs#L279-L295) using `predicate_must_hold_modulo_regions` while considering removal of the borrow. This includes requirements beyond the changed predicate argument itself.

`Option::filter` already had const-related `FnOnce`/`Destruct` bounds in [1.95.0](https://github.com/rust-lang/rust/blob/1.95.0/library/core/src/option.rs#L1578) as well as [1.98.0](https://github.com/rust-lang/rust/blob/1.98.0/library/core/src/option.rs#L1574), so it becoming const is **not** the change between these versions.

However, `Vec`'s destructor changes from [non-const in 1.95.0](https://github.com/rust-lang/rust/blob/1.95.0/library/alloc/src/vec/mod.rs#L4218) to [const in 1.98.0](https://github.com/rust-lang/rust/blob/1.98.0/library/alloc/src/vec/mod.rs#L4294), with `T: [const] Destruct` and allocator bounds. The [host-effect evaluator](https://github.com/rust-lang/rust/blob/1.98.0/compiler/rustc_trait_selection/src/traits/effects.rs#L459-L479) recursively examines fields for ADT destruction and handles non-const versus const destructors differently.

This suggests the lint's speculative const-destruction query may now recurse through `Value -> Vec<Value> -> Value`, where the older non-const destructor would stop evaluation. This is a source-supported hypothesis, not proof of the precise cycle or its complexity. No compiler-commit bisection has been performed.

### Version

```text
rustc 1.98.1 (48a229cea 2026-09-01)
binary: rustc
commit-hash: 48a229ceaefd4985c50990b14116b6d856af0985
commit-date: 2026-09-01
host: aarch64-apple-darwin
release: 1.98.1
LLVM version: 22.1.8

clippy 0.1.98 (48a229ceae 2026-09-01)
```

Passing comparison: `clippy 0.1.95 (59807616e1 2026-04-14)`.

The regression is confirmed between the tested 1.95.0 and 1.98.1 endpoints; intervening releases and a current nightly have not been tested. There is no ICE output or panic backtrace from the minimal reproducer.

**Comment by Gri-ffin** [comment]
I can't reproduce this on the current nightly `1.100.0-nightly (2026-09-09 a36d05efab632d1ddf90)`

**Comment by CommanderStorm** [comment]
Yea.. beta is worse though..

```
rustup run beta clippy-driver --edition=2024 repro.rs \
  --crate-type lib --emit metadata -o repro.rmeta \
  -W clippy::needless_borrows_for_generic_args

thread 'rustc' (1365002) has overflowed its stack
fatal runtime error: stack overflow, aborting
Aborted (core dumped)
```

Given that something has been fixed somewhere, I suspect there is something where backporting might be possible/nessary to prevent the ICE (?). 
I think I am going to start the bisection to where it broke/was fixed

@rustbot label beta-nominated

**Comment by CommanderStorm** [comment]
uhh..
that seems like something that I am not sure will be possible to fix beta for (?), or at least not easily 👀...
And I am also not sure that this is caused by something that clippy can actually fix..

This seems to have been ***FIXED*** by
- https://github.com/rust-lang/rust/pull/160619

`cargo bisect-rustc --start=2026-08-01 --end=2026-08-30 -c clippy --regress=success --timeout 5 -- clippy`

```
searched nightlies: from nightly-2026-08-01 to nightly-2026-08-30
regressed nightly: nightly-2026-08-22
searched commit range: https://github.com/rust-lang/rust/compare/8925ea358a0f265ca61026aadc7ecc506c545cbe...c656540d6467dee1381f0cbd882412d6bd1cd5ae
regressed commit: https://github.com/rust-lang/rust/commit/526c36b4c477994b7905ec6a084d6150e49076f5
```

The commit that seems to ***BREAK*** it is similarly not super clear (plus ages ago):

`cargo bisect-rustc --start=2026-06-01 --end=2026-07-30 -c clippy --timeout 5 -- clippy`

- https://github.com/rust-lang/rust/pull/157991


CC @JonathanBrouwer

```
searched nightlies: from nightly-2026-06-01 to nightly-2026-07-30
regressed nightly: nightly-2026-06-17
searched commit range: https://github.com/rust-lang/rust/compare/01dfd79246f1b2d5f146616deff08223a840a9ae...9e2abe0c6ab27fcbb95c30695188a75776e2feb1
regressed commit: https://github.com/rust-lang/rust/commit/9e2abe0c6ab27fcbb95c30695188a75776e2feb1
```

**Comment by Jarcho** [comment]
Given the stack overflow it's getting stuck in the resolver so it might be a rustc bug causing it, but it's possible we're passing a nonsense predicate.

**Comment by pfugate** [comment]
Bisected it one step further: **rust-lang/rust#160619 did not fix this, it changed solvers.** The old trait solver still loops on today's nightly, and the new solver sails through on stable *and* beta. Same reproducer, same lint, one flag:

| Toolchain | Solver | Result |
|---|---|---|
| stable `1.98.0 (88d9e12ae)` | old (default) | OOM-killed at a 4 GB cgroup cap after 9.3 s; uncapped it reached **22 GB RSS in ~70 s** |
| stable `1.98.0` | new (`RUSTC_BOOTSTRAP=1 -Znext-solver=globally`) | exit 0, 0.1 s |
| beta `1.99.0-beta.5 (fba1b05a7)` | old (default) | `thread 'rustc' has overflowed its stack`, 0.09 s |
| beta `1.99.0-beta.5` | new (`-Znext-solver=globally`) | exit 0 |
| nightly `1.100.0 (0fc141305 2026-09-11)` | new (default since #160619) | exit 0, 0.14 s |
| nightly `1.100.0` | old (`-Znext-solver=coherence`) | `thread 'rustc' has overflowed its stack`, 0.11 s |

So every toolchain whose non-coherence solving is the old solver dies, and every toolchain on the new solver passes. Unless the new solver reaches stable in 1.100, stable users keep hitting this after this cycle too.

**Why stable eats memory instead of overflowing.** On 1.98 the recursion is the same, but rustc grows that stack on the heap (`stacker`), so it presents as a memory runaway, not a crash. With `prlimit --as` the driver finally aborts and shows the loop, ~3,150,000 frames of:

```
clippy_lints::needless_borrows_for_generic_args::needless_borrow_count
  -> SelectionContext::evaluate_predicate_recursively
  -> InferCtxt::enter_forall::<HostEffectPredicate<TyCtxt>, ...>
  -> SelectionContext::evaluate_predicate_recursively
  ...
memory allocation of 939524096 bytes failed
```

That is the old solver's *evaluate* path, which `predicate_must_hold_modulo_regions` uses, descending through the host-effect predicate with nothing bounding it.

**Is the predicate nonsense?** (@Jarcho's question.) Take clippy out entirely and ask rustc the same question the lint asks, on either solver:

```rust
#![feature(const_trait_impl, const_destruct)]
use std::marker::Destruct;

pub enum Value {
    Array(Vec<Value>),
    Null,
}

const fn consume<T: [const] Destruct>(_t: T) {}

pub const fn go(v: Value) {
    consume(v)
}
```

```
error[E0275]: overflow evaluating the requirement `Value: [const] Destruct`
```

in 0.05 s, on both `-Znext-solver=globally` and `-Znext-solver=coherence`. The const-`Option::filter` form (`#![feature(const_option_ops)]`, `value.filter(keep)` with a `const fn keep`) gives the identical E0275. So the predicate is a real one that rustc itself answers with *overflow*: `Value: [const] Destruct` needs `Vec<Value>: [const] Destruct` needs `Value: [const] Destruct`. The new solver and rustc's normal fulfillment path both bound that and report it; the old solver's evaluate path does not, and the lint runs straight into it.

That also confirms @roji's hypothesis down to the commit: the bisected rollup rust-lang/rust#157991 carries rust-lang/rust#155616, whose description says *"this makes `Vec` `const Destruct` along the way."* Before it, `Vec<Value>` had no const destructor and the walk stopped at the first field.

**Where that leaves a backport.** The solver side is not backportable (agreed with @CommanderStorm). But the lint side is small: the lint is evaluating callee predicates that include `[const] Destruct` / `~const FnOnce` host-effect bounds for a borrow-removal it is only *considering*. Skipping host-effect predicates in that evaluation, or evaluating them in a way that treats overflow as "do not lint" (the answer the new solver effectively gives), keeps the lint sound and turns a compiler OOM into a missed suggestion on a recursive type. That would fit a 1.99 backport.

Two smaller things, each measured under the 4 GB cap on 1.98:

- A module-level `#![allow(clippy::needless_borrows_for_generic_args)]` in every module still dies; only the crate-root `allow` (or `-A`) skips the pass. Worth knowing for anyone hunting this in a workspace.
- The recursion in the *type* is what matters: `Array(Vec<u32>)` passes; `Array(Vec<Value>)` dies; derives make no difference.

Reproduction for the table (Linux, cgroup cap so the box survives):

```sh
systemd-run --user --scope -q -p MemoryMax=4G -p MemorySwapMax=0 \
  rustup run stable clippy-driver --edition=2021 repro.rs \
  --crate-type lib --emit metadata -o repro.rmeta \
  -W clippy::needless_borrows_for_generic_args
# add: env RUSTC_BOOTSTRAP=1 ... -Znext-solver=globally   → passes
# nightly, add: -Znext-solver=coherence                    → stack overflow
```

**Comment by flip1995** [comment]
beta-nominated label is also applied to the fix PR: https://github.com/rust-lang/rust-clippy/pull/17731. I will backport that fix.

**Comment by CommanderStorm** [comment]
thanks ❤️
